src/oauth2/controller/authorize.ts

```diff
@@ -190,11 +212,20 @@ class AuthorizeController extends Controller {
 
   }
 
-  async codeRedirect(ctx: Context, oauth2Client: OAuth2Client, redirectUri: string, state: string|undefined) {
+  async codeRedirect(
+    ctx: Context,
+    oauth2Client: OAuth2Client,
+    redirectUri: string,
+    state: string|undefined,
+    codeChallenge: string|undefined,
+    codeChallengeMethod: string|undefined,
```

Comment:
Could this type be refined to `'S256' | 'plain' | undefined` vs just `string` ?

Reply:
Is that valid syntax? I can't find anything in the TS docs about doing that. When I try it, I get an error on the calling side:
```
Argument of type 'string' is not assignable to parameter of type '"plain" | "S256"'.
```

Reply:
Got it. Combo of string literals
https://www.typescriptlang.org/docs/handbook/advanced-types.html#string-literal-types
and type assertions
https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions

Reply:
Basically, we want the strictest possible types for everything.

To fix the error, you also need to type the original definition the same way:

   let codeChallengeMethod: 'plain' | 'S256' | undefined;

It might also help to create a new reuable type for the enum. Ideally we use this everywhere this is applicable.

    type CodeChallengeMethod = 'plain' | 'S256';

The benefit if taking this approach and validating values *before* they enter the function, is that the function can only ever be called with a correct value, which means no validation in the function is needed.

I wrote a bit about this here:

https://evertpot.com/typescript-changes-code/

The official docs are here:

https://www.typescriptlang.org/docs/handbook/advanced-types.html#string-literal-types

Reply:
The validation still needs to happen _somewhere_ since these are user provided values. Even though TS will now prevent us from calling this function with a different value, validation now just needs to happen before calling the function.